modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java

```diff
@@ -622,6 +631,24 @@ private boolean isBaselineSatisfied(BaselineTopology blt, List<ClusterNode> srvN
     @Override public void onStateFinishMessage(ChangeGlobalStateFinishMessage msg) {
         DiscoveryDataClusterState discoClusterState = globalState;
 
+        if (ctx.clientNode() && !ctx.isDaemon() && changeStateInProgress != null) {
```

Comment:
I don't like this style when we inline a very specific logic as plain blocks of code into existing methods.

Why not to create a new class called for instance ClientStateChangeSynchronizer and encapsulate this logic there (including aforementioned ConcurrentHashMap)?

Reply:
Done